File: 2023/solutions/day12.py
from input import Input
from functools import cache

def parse_input(input: Input) -> list[tuple[str, list[int]]]:
    return [(l, [int(n) for n in r.split(",")]) for l, r in [line.split() for line in input.lines]]

@cache
def valid_combinations(springs: str, groups: tuple[int, ...]) -> int:
    springs = springs.lstrip(".")
    if len(groups) == 0:
        return "#" not in springs
    if len(springs) == 0:
        return len(groups) == 0
    
    if springs[0] == "?":
        return valid_combinations("." + springs[1:], groups) + valid_combinations("#" + springs[1:], groups)
    elif springs[0] == "#":
        remaining = groups[0]
        if "." not in springs[:remaining]:
            if remaining < len(springs) and springs[remaining] != "#":
                return valid_combinations(springs[remaining + 1:], groups[1:])
            elif len(springs) == remaining:
                return len(groups) == 1
    return 0

def solve_first(input: Input) -> object:
    rows = parse_input(input)
    return sum(valid_combinations(row[0], tuple(row[1])) for row in rows)

def solve_second(input: Input) -> object:
    rows = parse_input(input)
    rows = [("?".join(springs for _ in range(5)), groups * 5) for springs, groups in rows]
    return sum(valid_combinations(row[0], tuple(row[1])) for row in rows)

File: 2025/src/day12.fs
module Day12
open Input
open Common

[<Literal>]
let name = "Christmas Tree Farm"

let parseInput (input: Input) = 
    let lastIdx = input.lines |> List.findIndexBack ((=) "")
    let shapes = 
        input.lines.[.. (lastIdx - 1)]
        |> List.split ((=) "")
        |> List.map (fun lines -> lines.[1..] |> Parsing.map2d Vec2.ofTpl id |> List.ofSeq)
    let areas = 
        input.lines.[(lastIdx + 1) ..]
        |> List.map (fun line -> 
            let nums = line |> Parsing.ints
            Vec2(nums.[0], nums.[1]), nums.[2..]
        )
    shapes, areas

let shapeArea (shape: (Vec2 * char) list) = 
    let width = shape |> List.map (fst >> _.x) |> List.max |> (+) 1
    let height = shape |> List.map (fst >> _.y) |> List.max |> (+) 1
    width * height

let canFit (shapes: (Vec2 * char) list list) (area: (Vec2 * int list)) =
    let areaSize = area |> fst |> fun v -> v.x * v.y
    let requiredSize = area |> snd |> List.mapi (fun i n -> n * (shapeArea shapes.[i])) |> List.sum
    requiredSize <= areaSize

let solveFirst (input: Input) = 
    let (shapes, areas) = parseInput input
    areas
    |> List.filter (canFit shapes)
    |> List.length

let solveSecond (input: Input) = 
    // Day 12 only has 1 puzzle
    null :> obj
